powdr/src/lib.rs

```diff
@@ -12,3 +12,179 @@ pub use powdr_pipeline::Pipeline;
 pub use powdr_number::Bn254Field;
 pub use powdr_number::FieldElement;
 pub use powdr_number::GoldilocksField;
+
+use riscv::CompilerOptions;
+
+use std::fs::File;
+use std::path::Path;
+use std::path::PathBuf;
+use std::time::Instant;
+
+pub struct Session {
+    pipeline: Pipeline<GoldilocksField>,
+    out_path: String,
+}
+
+const DEFAULT_PKEY: &str = "pkey.bin";
+const DEFAULT_VKEY: &str = "vkey.bin";
+
+impl Session {
+    pub fn new(guest_path: &str, out_path: &str) -> Self {
+        Session {
+            pipeline: pipeline_from_guest(guest_path, Path::new(out_path)),
+            out_path: out_path.into(),
+        }
+        .with_backend(powdr_backend::BackendType::Plonky3)
+    }
+
+    pub fn into_pipeline(self) -> Pipeline<GoldilocksField> {
+        self.pipeline
+    }
+
+    pub fn pipeline(&self) -> &Pipeline<GoldilocksField> {
+        &self.pipeline
+    }
+
+    pub fn with_backend(self, backend: backend::BackendType) -> Self {
+        Session {
+            pipeline: self.pipeline.with_backend(backend, None),
+            ..self
+        }
+    }
+
+    pub fn write<S: serde::Serialize>(self, channel: u32, data: &S) -> Self {
+        Session {
+            pipeline: self.pipeline.add_data(channel, data),
```

Comment:
Does it make sense to have two different terms for session and pipeline? Why not `write` for both?

Reply:
I think `add_data` is more accurate which then made sense for Pipeline, whereas since Session is user facing, I think something more "syscall-like" would be better, like write/read etc (which we also have in the runtime syscalls).